// @trigger.dev/core@3.3.5/dist/esm/v3/runMetadata/index.js
import { getGlobal, registerGlobal } from "../utils/globals.js";
import { NoopRunMetadataManager } from "./noopManager.js";
const API_NAME = "run-metadata";
const NOOP_MANAGER = new NoopRunMetadataManager();
export class RunMetadataAPI {
    static _instance;
    constructor() { }
    static getInstance() {
        if (!this._instance) {
            this._instance = new RunMetadataAPI();
        }
        return this._instance;
    }
    setGlobalManager(manager) {
        return registerGlobal(API_NAME, manager);
    }
    #getManager() {
        return getGlobal(API_NAME) ?? NOOP_MANAGER;
    }
    enterWithMetadata(metadata) {
        this.#getManager().enterWithMetadata(metadata);
    }
    current() {
        return this.#getManager().current();
    }
    getKey(key) {
        return this.#getManager().getKey(key);
    }
    setKey(key, value) {
        return this.#getManager().setKey(key, value);
    }
    deleteKey(key) {
        return this.#getManager().deleteKey(key);
    }
    incrementKey(key, value) {
        return this.#getManager().incrementKey(key, value);
    }
    decrementKey(key, value) {
        return this.#getManager().decrementKey(key, value);
    }
    appendKey(key, value) {
        this.#getManager().appendKey(key, value);
    }
    removeFromKey(key, value) {
        this.#getManager().removeFromKey(key, value);
    }
    update(metadata) {
        return this.#getManager().update(metadata);
    }
    stream(key, value, signal) {
        return this.#getManager().stream(key, value, signal);
    }
    flush(requestOptions) {
        return this.#getManager().flush(requestOptions);
    }
}
